craft_application/services/project.py

```diff
@@ -0,0 +1,362 @@
+            raise errors.ProjectDirectoryTypeError(self._project_dir)
+        try:
+            path = (self._project_dir / f"{self._app.name}.yaml").resolve(strict=True)
+        except FileNotFoundError as err:
+            raise errors.ProjectFileMissingError(
+                f"Project file '{self._app.name}.yaml' not found in '{self._project_dir}'.",
+                details="The project file could not be found.",
+                resolution="Ensure the project file exists.",
+                retcode=os.EX_NOINPUT,
+            ) from err
+        emit.trace(f"Project file found at {path}")
+        return path
+
+    @final
+    def _load_raw_project(self) -> dict[str, Any]:
+        """Get the raw project data structure.
+
+        This loads the project file from the given path, parses the YAML, and returns
+        that raw data structure. This method should be used with care, as the project
+        does not have any preprocessors applied.
+        """
+        if self.__raw_project:
+            return self.__raw_project
+        project_path = self.resolve_project_file_path()
+        with project_path.open() as project_file:
+            emit.debug(f"Loading project file '{project_path!s}")
+            raw_yaml = util.safe_yaml_load(project_file)
+        if not isinstance(raw_yaml, dict):
+            raise errors.ProjectFileInvalidError(raw_yaml)
+        self.__raw_project = cast(dict[str, Any], raw_yaml)
+        return self.__raw_project
+
+    def _app_preprocess_project(
+        self,
+        project: dict[str, Any],
+        *,
+        build_on: str,
+        build_for: str,
+        platform: str,
+    ) -> None:
+        """Run any application-specific pre-processing on the project, in-place.
+
+        This includes any application-specific transformations on a project's raw data
+        structure before it gets rendered as a pydantic model. Some examples of
+        processing to do here include:
+
+        - Applying extensions
+        - Adding "hidden" app-specific parts
+        - Processing grammar for keys other than parts
+        """
+
+    def _app_render_legacy_platforms(self) -> dict[str, craft_platforms.PlatformDict]:
+        """Application-specific rendering function if no platforms are declared.
+
+        This method is intended to be overridden by an application-specific service
+        if the root project has no ``platforms`` key. In this case, it should return
+        a dictionary structured the same way as the ``platforms`` key in a project
+        for use in that project. This dictionary is less strict than what is expected
+        in the project file's ``platforms`` key. For example, the ``build-for``
+        key for each platform **MAY** contain multiple targets.
+        """
+        raise errors.CraftValidationError(
+            f"{self._app.name}.yaml must contain a 'platforms' key."
+        )
+
+    @final
+    def get_platforms(self) -> dict[str, craft_platforms.PlatformDict]:
```

Comment:
how come this method is public? Who, do you figure, will need to use this? Other than tests I mean

Reply:
The BuildPlanService uses it to generate the build plan. This is also the primary reason for having `_app_render_legacy_platforms()` - the BuildPlanService expects to get a `platforms` entry that craft-platforms can use to generate the exhaustive build plan.